File: src/modules/time/reminders.ts
import { Temporal } from '@js-temporal/polyfill';
import { validate as validUuid } from 'uuid';
import { z } from 'zod';
import { civilDateSchema, localTimeSchema, resolveTimeValue, timeValueSchema, timeZoneSchema, type TimeValue } from './values.js';
import { occurrenceKeySchema } from './recurrence.js';

const offset = z.number().int().min(0).max(10080);
const absoluteTimeSchema = z.object({ date: civilDateSchema, time: localTimeSchema, timeZone: timeZoneSchema }).strict();
export const reminderRuleSchema = z.discriminatedUnion('kind', [
  z.object({ kind: z.literal('before_start'), offsetMinutes: offset }).strict(),
  z.object({ kind: z.literal('before_deadline'), offsetMinutes: offset }).strict(),
  z.object({ kind: z.literal('on_scheduled_day_at'), localTime: localTimeSchema }).strict(),
  z.object({ kind: z.literal('on_deadline_day_at'), localTime: localTimeSchema }).strict(),
  z.object({ kind: z.literal('absolute'), absolute: absoluteTimeSchema }).strict(),
]);
export type ReminderRule = z.infer<typeof reminderRuleSchema>;
export type ReminderTrigger = { state: 'active'; triggerAt: string } | { state: 'inactive_base_missing'; triggerAt: null };

export function notificationId(reminderId: string, occurrenceKey: string | null = null): string {
  if (!validUuid(reminderId)) throw new RangeError('Invalid reminder UUID');
  if (occurrenceKey !== null) occurrenceKeySchema.parse(occurrenceKey);
  return `r:${reminderId.toLowerCase()}:${occurrenceKey ?? 'once'}`;
}

/** Caller supplies the effective occurrence schedule (including its durable override). */
export function resolveReminderTrigger(input: { reminder: ReminderRule; schedule: TimeValue | null; deadline: TimeValue | null; deviceTimeZone: string; isRecurring?: boolean; occurrenceKey?: string | null }): ReminderTrigger {
  const reminder = reminderRuleSchema.parse(input.reminder);
  timeZoneSchema.parse(input.deviceTimeZone);
  if (input.occurrenceKey != null) occurrenceKeySchema.parse(input.occurrenceKey);
  if (reminder.kind === 'absolute') {
    if (input.isRecurring === true && input.occurrenceKey == null) throw new RangeError('ABSOLUTE_REMINDER_REQUIRES_OCCURRENCE');
    const resolved = resolveTimeValue(reminder.absolute);
    if (resolved.kind !== 'timed') throw new Error('Absolute reminder requires a time');
    return { state: 'active', triggerAt: resolved.instant };
  }
  const base = reminder.kind === 'before_start' || reminder.kind === 'on_scheduled_day_at' ? input.schedule : input.deadline;
  if (base === null) return { state: 'inactive_base_missing', triggerAt: null };
  const normalized = timeValueSchema.parse(base);
  if (reminder.kind === 'before_start' || reminder.kind === 'before_deadline') {
    if (normalized.time === null) return { state: 'inactive_base_missing', triggerAt: null };
    const resolved = resolveTimeValue(normalized);
    if (resolved.kind !== 'timed') throw new Error('Timed base expected');
    return { state: 'active', triggerAt: Temporal.Instant.from(resolved.instant).subtract({ minutes: reminder.offsetMinutes }).toString() };
  }
  if (normalized.time !== null) return { state: 'inactive_base_missing', triggerAt: null };
  const resolved = resolveTimeValue({ date: normalized.date, time: reminder.localTime, timeZone: input.deviceTimeZone });
  if (resolved.kind !== 'timed') throw new Error('Timed reminder expected');
  return { state: 'active', triggerAt: resolved.instant };
}

export const REMINDER_WINDOW_DAYS = 14;
export const MAX_PENDING_REMINDERS = 50;
export interface ReminderCandidate {
  reminderId: string;
  occurrenceKey: string | null;
  trigger: ReminderTrigger;
  /** false for a deleted reminder or a completed/skipped/deleted task or occurrence. */
  eligible: boolean;
}
export interface ScheduledRequest { notificationId: string; triggerAt: string }
export interface SchedulingReceipt extends ScheduledRequest { acceptedAt: string }
export type ReminderProgrammingState = 'needs_scheduling' | 'scheduled' | 'pending_window' | 'pending_capacity' | 'notifications_disabled' | 'inactive_base_missing' | 'removed' | 'missed' | 'display_unknown';

/** Pure desired-state plan. Only actual system pending requests prove scheduling. */
export function planReminderReconciliation(input: {
  referenceInstant: string;
  deviceTimeZone: string;
  notificationsAuthorized: boolean;
  candidates: readonly ReminderCandidate[];
  systemPending?: readonly ScheduledRequest[];
  schedulingReceipts?: readonly SchedulingReceipt[];
}): { desired: ScheduledRequest[]; addOrReplace: ScheduledRequest[]; remove: string[]; states: Record<string, ReminderProgrammingState> } {
  timeZoneSchema.parse(input.deviceTimeZone);
  const now = Temporal.Instant.from(input.referenceInstant);
  const horizon = now.toZonedDateTimeISO(input.deviceTimeZone).add({ days: REMINDER_WINDOW_DAYS }).toInstant();
  const pending = new Map((input.systemPending ?? []).map((request) => [request.notificationId, Temporal.Instant.from(request.triggerAt).toString()]));
  const receipts = input.schedulingReceipts ?? [];
  const states: Record<string, ReminderProgrammingState> = Object.create(null) as Record<string, ReminderProgrammingState>;
  const eligible: ScheduledRequest[] = [];
  for (const candidate of input.candidates) {
    const id = notificationId(candidate.reminderId, candidate.occurrenceKey);
    if (Object.hasOwn(states, id)) throw new RangeError('Duplicate notification candidate');
    if (!candidate.eligible) { states[id] = 'removed'; continue; }
    if (candidate.trigger.state === 'inactive_base_missing') { states[id] = 'inactive_base_missing'; continue; }
    const trigger = Temporal.Instant.from(candidate.trigger.triggerAt);
    if (Temporal.Instant.compare(trigger, now) <= 0) {
      const wasScheduled = receipts.some((receipt) => receipt.notificationId === id && Temporal.Instant.compare(receipt.triggerAt, trigger) === 0 && Temporal.Instant.compare(receipt.acceptedAt, trigger) < 0);
      states[id] = wasScheduled ? 'display_unknown' : 'missed';
      continue;
    }
    if (!input.notificationsAuthorized) { states[id] = 'notifications_disabled'; continue; }
    if (Temporal.Instant.compare(trigger, horizon) > 0) { states[id] = 'pending_window'; continue; }
    states[id] = 'pending_capacity';
    eligible.push({ notificationId: id, triggerAt: trigger.toString() });
  }
  eligible.sort((a, b) => Temporal.Instant.compare(a.triggerAt, b.triggerAt) || a.notificationId.localeCompare(b.notificationId, 'en'));
  const desired = eligible.slice(0, MAX_PENDING_REMINDERS);
  const desiredIds = new Set(desired.map((request) => request.notificationId));
  const addOrReplace: ScheduledRequest[] = [];
  for (const request of desired) {
    const accepted = pending.get(request.notificationId) === request.triggerAt;
    states[request.notificationId] = accepted ? 'scheduled' : 'needs_scheduling';
    if (!accepted) addOrReplace.push(request);
  }
  const remove = [...pending.keys()].filter((id) => id.startsWith('r:') && !desiredIds.has(id)).sort();
  return { desired, addOrReplace, remove, states };
}

File: ios/Planner/Domain/Time/Reminders.swift
import Foundation

/// One reminder rule (04_Backend/03_Data_Model.md §3): relative to the planned time or the deadline,
/// at a time on the planned or due day, or absolute.
nonisolated enum ReminderRule: Hashable, Sendable {
    case beforeStart(minutes: Int)
    case beforeDeadline(minutes: Int)
    case onScheduledDay(LocalTime)
    case onDeadlineDay(LocalTime)
    case absolute(TimeValue)

    init?(kind: String?, offsetMinutes: Int?, localTime: String?, absoluteDate: String?, absoluteTime: String?, absoluteZone: String?) {
        switch kind {
        case "before_start": guard let offsetMinutes else { return nil }; self = .beforeStart(minutes: offsetMinutes)
        case "before_deadline": guard let offsetMinutes else { return nil }; self = .beforeDeadline(minutes: offsetMinutes)
        case "on_scheduled_day_at": guard let time = LocalTime(localTime) else { return nil }; self = .onScheduledDay(time)
        case "on_deadline_day_at": guard let time = LocalTime(localTime) else { return nil }; self = .onDeadlineDay(time)
        case "absolute":
            guard let value = TimeValue(date: absoluteDate, time: absoluteTime, timeZone: absoluteZone), value.time != nil else { return nil }
            self = .absolute(value)
        default: return nil
        }
    }

    var kind: String {
        switch self {
        case .beforeStart: "before_start"
        case .beforeDeadline: "before_deadline"
        case .onScheduledDay: "on_scheduled_day_at"
        case .onDeadlineDay: "on_deadline_day_at"
        case .absolute: "absolute"
        }
    }

    /// `rule` of `reminder.set`.
    var payload: JSONPayload {
        switch self {
        case .beforeStart(let minutes): ["kind": "before_start", "offsetMinutes": .int(minutes)]
        case .beforeDeadline(let minutes): ["kind": "before_deadline", "offsetMinutes": .int(minutes)]
        case .onScheduledDay(let time): ["kind": "on_scheduled_day_at", "localTime": .string(time.description)]
        case .onDeadlineDay(let time): ["kind": "on_deadline_day_at", "localTime": .string(time.description)]
        case .absolute(let value): ["kind": "absolute", "absolute": value.payload]
        }
    }

    /// Whether the rule needs the planned value or the deadline.
    var usesDeadline: Bool {
        switch self {
        case .beforeDeadline, .onDeadlineDay: true
        default: false
        }
    }

    /// "30 min avant", "À l'heure prévue", "À 09:00 le jour prévu".
    var label: String {
        switch self {
        case .beforeStart(0): "À l’heure prévue"
        case .beforeStart(let minutes): Self.offsetText(minutes) + " avant"
        case .beforeDeadline(0): "À l’heure de l’échéance"
        case .beforeDeadline(let minutes): Self.offsetText(minutes) + " avant l’échéance"
        case .onScheduledDay(let time): "À \(time) le jour prévu"
        case .onDeadlineDay(let time): "À \(time) le jour de l’échéance"
        case .absolute(let value): "Le " + value.date.noon().formatted(.dateTime.day().month(.abbreviated)) + " à \(value.time?.description ?? "")"
        }
    }

    private static func offsetText(_ minutes: Int) -> String {
        if minutes % 1440 == 0 { return minutes == 1440 ? "1 jour" : "\(minutes / 1440) jours" }
        return DurationText.format(minutes)
    }
}

nonisolated enum ReminderTrigger: Equatable, Sendable {
    case active(Date)
    case baseMissing
}

nonisolated enum ReminderMath {
    static let windowDays = 14
    static let maxPending = 50

    /// `r:<reminderId>:<occurrenceKey|once>` (03_iOS/03_Notifications_EventKit_Widgets.md §1.3).
    static func notificationId(reminderId: String, occurrenceKey: String?) -> String {
        "r:\(reminderId.lowercased()):\(occurrenceKey ?? "once")"
    }

    /// The caller passes the effective schedule of the occurrence (its override included).
    static func trigger(_ rule: ReminderRule, schedule: TimeValue?, deadline: TimeValue?, deviceZone: TimeZone) -> ReminderTrigger {
        switch rule {
        case .absolute(let value):
            return value.instant.map(ReminderTrigger.active) ?? .baseMissing
        case .beforeStart(let minutes), .beforeDeadline(let minutes):
            let base = rule.usesDeadline ? deadline : schedule
            guard let base, base.time != nil, let instant = base.instant else { return .baseMissing }
            return .active(instant.addingTimeInterval(-TimeInterval(minutes * 60)))
        case .onScheduledDay(let time), .onDeadlineDay(let time):
            let base = rule.usesDeadline ? deadline : schedule
            guard let base, base.time == nil,
                  let resolved = TimeResolver.resolve(date: base.date, time: time, zone: deviceZone.identifier) else { return .baseMissing }
            return .active(resolved.instant)
        }
    }

    nonisolated enum State: String, Hashable, Sendable {
        case needsScheduling = "needs_scheduling"
        case scheduled
        case pendingWindow = "pending_window"
        case pendingCapacity = "pending_capacity"
        case notificationsDisabled = "notifications_disabled"
        case baseMissing = "inactive_base_missing"
        case removed
        case missed
        case displayUnknown = "display_unknown"
    }

    nonisolated struct Candidate: Sendable {
        let notificationId: String
        let trigger: ReminderTrigger
        /// False for a deleted reminder or a completed, skipped or deleted task or occurrence.
        let eligible: Bool
    }

    nonisolated struct Request: Hashable, Sendable {
        let notificationId: String
        let triggerAt: Date
    }

    nonisolated struct Plan: Sendable {
        let desired: [Request]
        let addOrReplace: [Request]
        let remove: [String]
        let states: [String: State]
    }

    /// Pure desired state (ported from reminders.ts): only requests actually pending in iOS prove scheduling;
    /// an acceptance recorded before a past trigger proves programming, never display.
    static func plan(now: Date, zone: TimeZone, authorized: Bool, candidates: [Candidate],
                     systemPending: [Request], acceptances: [String: (triggerAt: Date, acceptedAt: Date)]) -> Plan {
        let local = TimeResolver.project(now, into: zone)
        let second = Int(now.timeIntervalSince1970.rounded(.down)) % 60
        let horizonWall = WallClock.seconds(local.date.adding(days: windowDays), local.time, second: second)
        let horizon = TimeResolver.instant(wallSeconds: horizonWall, zone: zone).addingTimeInterval(now.timeIntervalSince1970 - now.timeIntervalSince1970.rounded(.down))
        let pending = Dictionary(systemPending.map { ($0.notificationId, $0.triggerAt) }, uniquingKeysWith: { first, _ in first })
        var states: [String: State] = [:]
        var eligible: [Request] = []
        for candidate in candidates {
            let id = candidate.notificationId
            guard candidate.eligible else { states[id] = .removed; continue }
            guard case .active(let trigger) = candidate.trigger else { states[id] = .baseMissing; continue }
            if trigger <= now {
                if let accepted = acceptances[id], accepted.triggerAt == trigger, accepted.acceptedAt < trigger {
                    states[id] = .displayUnknown
                } else {
                    states[id] = .missed
                }
                continue
            }
            guard authorized else { states[id] = .notificationsDisabled; continue }
            guard trigger <= horizon else { states[id] = .pendingWindow; continue }
            states[id] = .pendingCapacity
            eligible.append(Request(notificationId: id, triggerAt: trigger))
        }
        eligible.sort { ($0.triggerAt, $0.notificationId) < ($1.triggerAt, $1.notificationId) }
        let desired = Array(eligible.prefix(maxPending))
        let desiredIds = Set(desired.map(\.notificationId))
        var addOrReplace: [Request] = []
        for request in desired {
            let accepted = pending[request.notificationId].map { abs($0.timeIntervalSince(request.triggerAt)) < 0.5 } ?? false
            states[request.notificationId] = accepted ? .scheduled : .needsScheduling
            if !accepted { addOrReplace.append(request) }
        }
        let remove = pending.keys.filter { $0.hasPrefix("r:") && !desiredIds.contains($0) }.sorted()
        return Plan(desired: desired, addOrReplace: addOrReplace, remove: remove, states: states)
    }
}
